examples/blockchain/src/network.rs

```diff
@@ -233,130 +95,200 @@ impl aleph_bft::Network<NetworkData> for Network {
+        let mut dns_hello_ticker = Delay::new(dns_hello_ticker_delay).fuse();
+
         loop {
-            futures::select! {
+            let mut buffer = Vec::new();
+            tokio::select! {
+
+                event = self.listener.accept() => match event {
+                    Ok((mut socket, _addr)) => {
+                        match socket.read_to_end(&mut buffer).await {
+                            Ok(_) => {
+                                let message = Message::decode(&mut &buffer[..]);
+                                debug!("Received message: {:?}", message);
+                                match message {
+                                    Ok(Message::Consensus(data)) => self.consensus_tx.unbounded_send(data).expect("Network must listen"),
+                                    Ok(Message::Block(block)) => {
+                                        debug!(target: "Blockchain-network", "Received block num {:?}", block.num);
+                                        self.block_tx
+                                            .unbounded_send(block)
+                                            .expect("Blockchain process must listen");
+                                    },
+                                    Ok(Message::DNSHello(id, address)) => {
+                                        self.addresses[id as usize] = Some(address);
+                                    },
+                                    Ok(Message::DNSRequest(id, address)) => self.dns_response(id as usize, address),
+                                    Ok(Message::DNSResponse(addresses)) => for (id, addr) in addresses.iter().enumerate() {
+                                        if let Some(addr) = addr {
+                                            self.addresses[id as usize] = Some(addr.clone());
+                                        };
+                                    },
+                                    Err(_) => (),
+                                };
+                            },
+                            Err(_) => {
+                                error!("Could not decode incoming data");
+                            },
+                        }
+                    },
+                    Err(e) => {
+                        error!("Couldn't accept connection: {:?}", e);
+                    },
+                },
+
+                _ = &mut dns_ticker => {
+                    if self.addresses.iter().any(|a| a.is_none()) {
+                        self.send(Message::DNSRequest(self.id as u32, self.address.clone()), Recipient::Everyone);
+                        debug!("Requesting IP addresses");
+                    }
+                    dns_ticker = Delay::new(dns_ticker_delay).fuse();
+                },
+
+                _ = &mut dns_hello_ticker => {
+                    self.send(Message::DNSHello(self.id as u32, self.address.clone()), Recipient::Everyone);
```

Comment:
I feel like right now Hello and Request are doing a lot of similar things. I think it is fine as it is in current state, but maybe gossiping Hello would be more like an actual network? Then you could also remove the Request sending.

Reply:
`Hello` is a quick fix for handling bootnode crashes, but I do agree that a small p2p module could be extracted from the current network implementation. Leaving as it is.